core/templates/services/exploration-improvements-task-registry.service.ts

```diff
@@ -250,64 +261,70 @@ export class ExplorationImprovementsTaskRegistryService {
           ...{target_id: newStateName},
         })
       ])),
-      oldStateTasks.supportingStateStats);
+      new SupportingStateStats(
+        this.expStats.getStateStats(newStateName),
+        oldStateTasks.supportingStats.answerStats,
+        oldStateTasks.supportingStats.cstPlaythroughIssues,
+        oldStateTasks.supportingStats.eqPlaythroughIssues,
+        oldStateTasks.supportingStats.misPlaythroughIssues));
 
     for (const newTask of newStateTasks) {
-      this.tasksByType.get(newTask.taskType).push(newTask);
+      if (newTask.isOpen()) {
+        this.pushOpenTask(newTask);
+      }
     }
     for (const oldTask of oldStateTasks) {
+      if (oldTask.isOpen()) {
+        this.popOpenTask(oldTask);
+      }
       oldTask.markAsObsolete();
     }
 
     this.tasksByState.set(newStateName, newStateTasks);
     // Map uses the reserved keyword delete as a method name.
     // eslint-disable-next-line dot-notation
     this.tasksByState.delete(oldStateName);
-    this.expStats = this.expStats.createNewWithStateRenamed(
-      oldStateName, newStateName);
   }
 
-  onChangeInteraction(stateName: string): void {
-    this.tasksByState.get(stateName).refresh(this.expStats, this.config);
+  onStateInteractionSaved(state: State): void {
+    this.refreshStateTasks(state.name);
   }
 
   getOpenHighBounceRateTasks(): HbrTask[] {
-    return <HbrTask[]> this.tasksByType.get(
-      ImprovementsConstants.TASK_TYPE_HIGH_BOUNCE_RATE)
-      .filter(t => t.isOpen());
+    return <HbrTask[]> this.openTasksByType.get(
+      ImprovementsConstants.TASK_TYPE_HIGH_BOUNCE_RATE);
   }
 
   getOpenIneffectiveFeedbackLoopTasks(): IflTask[] {
-    return <IflTask[]> this.tasksByType.get(
-      ImprovementsConstants.TASK_TYPE_INEFFECTIVE_FEEDBACK_LOOP)
-      .filter(t => t.isOpen());
+    return <IflTask[]> this.openTasksByType.get(
+      ImprovementsConstants.TASK_TYPE_INEFFECTIVE_FEEDBACK_LOOP);
   }
 
   getOpenNeedsGuidingResponsesTasks(): NgrTask[] {
-    return <NgrTask[]> this.tasksByType.get(
-      ImprovementsConstants.TASK_TYPE_NEEDS_GUIDING_RESPONSES)
-      .filter(t => t.isOpen());
+    return <NgrTask[]> this.openTasksByType.get(
+      ImprovementsConstants.TASK_TYPE_NEEDS_GUIDING_RESPONSES);
   }
 
   getOpenSuccessiveIncorrectAnswersTasks(): SiaTask[] {
-    return <SiaTask[]> this.tasksByType.get(
-      ImprovementsConstants.TASK_TYPE_SUCCESSIVE_INCORRECT_ANSWERS)
-      .filter(t => t.isOpen());
+    return <SiaTask[]> this.openTasksByType.get(
+      ImprovementsConstants.TASK_TYPE_SUCCESSIVE_INCORRECT_ANSWERS);
   }
 
-  getSupportingStateStats(task: ExplorationTask): SupportingStateStats {
-    if (!this.tasksByState.has(task.targetId)) {
-      throw new Error('Unregistered task has no supporting stats');
+  makeStateTasksObject(): {[stateName: string]: StateTasks} {
```

Comment:
Let's make a domain object if you want this to be a conceptual thing. No passing raw dicts around, especially if the keys can be fixed.